# Question 1:
<p style="font-size:22px;">Report 3 rules with support at least 0.2 and confidence at least 0.9. Specify for each of them the support and the confidence.</p>

In [1]:
import math
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules\


data = pd.read_csv(r"mammographic_masses.csv",delimiter=",",header=0)

#preprocessing: change to one hot encoding so as to be able to use apriori from mlxtend
d=data.values.tolist()


# NAN == "?"
for i in range(len(d)):
    j = 0
    while True:
        if d[i][j] == '?':  
            del d[i][j]     
            j -= 1          
        j += 1
        if j > len(d[i]) - 1:
            break


#adding attributes
for i in range(len(d)):
    for j in range (len(d[i])):
        d[i][j]=data.columns[j] + "=" +str(d[i][j])

            
te = TransactionEncoder()
te_ary = te.fit(d).transform(d)

df = pd.DataFrame(te_ary, columns=te.columns_)

#computing frequent itemsets and association rules
frequent_itemsets = apriori(df, min_support=0.2, use_colnames=True)

a=association_rules(frequent_itemsets, metric="confidence", min_threshold=0.9)

selected_rules = a[(a['support'] >= 0.2) & (a['confidence'] >= 0.9)]
selected_rules[["antecedents", "consequents", "support", "confidence"]]

antecedents   consequents   support  confidence
0                       (Severity=1)   (Density=3)  0.391259    0.933002
1              (Margin=1, Density=3)   (BI-RADS=4)  0.263267    0.900356
2             (Severity=0, Margin=1)   (BI-RADS=4)  0.269511    0.918440
3            (BI-RADS=5, Severity=1)   (Density=3)  0.277836    0.933566
4             (BI-RADS=5, Density=3)  (Severity=1)  0.277836    0.911263
5              (Severity=1, Shape=4)   (Density=3)  0.288241    0.929530
6   (BI-RADS=4, Margin=1, Density=3)  (Severity=0)  0.238293    0.905138
7  (Margin=1, Severity=0, Density=3)   (BI-RADS=4)  0.238293    0.927126
8   (Shape=4, BI-RADS=5, Severity=1)   (Density=3)  0.221644    0.930131
9    (BI-RADS=5, Shape=4, Density=3)  (Severity=1)  0.221644    0.914163

<p style="font-size:20px;">So, The table displays all rules with support >= 0.2 and confidence >= to 0.9. And the the rules along with their actual support and confidence are displayed.</p>

# Question 2
<p style="font-size:22px;">Mining the rules that can help us to predicting whether a given instance is benign or malign.</p>

In [2]:
import math
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules\


data = pd.read_csv(r"mammographic_masses.csv",delimiter=",",header=0)

#preprocessing: change to one hot encoding so as to be able to use apriori from mlxtend
d=data.values.tolist()


# NAN == "?"
for i in range(len(d)):
    j = 0
    while True:
        if d[i][j] == '?':  
            del d[i][j]     
            j -= 1          
        j += 1
        if j > len(d[i]) - 1:
            break


#adding attributes
for i in range(len(d)):
    for j in range (len(d[i])):
        d[i][j]=data.columns[j] + "=" +str(d[i][j])

            
te = TransactionEncoder()
te_ary = te.fit(d).transform(d)

df = pd.DataFrame(te_ary, columns=te.columns_)

#computing frequent itemsets and association rules
frequent_itemsets = apriori(df, min_support=0.1, use_colnames=True)

rule = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.9)

rule[["antecedents", "consequents", "support", "confidence"]]



antecedents   consequents   support  \
0                                    (Margin=5)   (Density=3)  0.126951   
1                                  (Severity=1)   (Density=3)  0.391259   
2                         (Margin=1, Density=3)   (BI-RADS=4)  0.263267   
3                        (Severity=0, Margin=1)   (BI-RADS=4)  0.269511   
4                         (Severity=0, Shape=1)   (BI-RADS=4)  0.151925   
5                         (Shape=2, Severity=0)   (BI-RADS=4)  0.140479   
6                       (BI-RADS=5, Severity=1)   (Density=3)  0.277836   
7                        (BI-RADS=5, Density=3)  (Severity=1)  0.277836   
8                        (Margin=4, Severity=1)   (Density=3)  0.169615   
9                        (Margin=5, Severity=1)   (Density=3)  0.106139   
10                          (Margin=5, Shape=4)   (Density=3)  0.111342   
11                        (Severity=0, Shape=1)   (Density=3)  0.151925   
12                        (Severity=1, Shape=4)   (Density=3)  0.288241   
13                        (Severity=0, Shape=1)    (Margin=1)  0.155047   
14             (BI-RADS=4, Margin=1, Density=3)  (Severity=0)  0.238293   
15            (Margin=1, Severity=0, Density=3)   (BI-RADS=4)  0.238293   
16              (BI-RADS=4, Shape=1, Density=3)    (Margin=1)  0.144641   
17               (Shape=1, Margin=1, Density=3)   (BI-RADS=4)  0.144641   
18              (BI-RADS=4, Shape=1, Density=3)  (Severity=0)  0.141519   
19             (BI-RADS=4, Severity=0, Shape=1)   (Density=3)  0.141519   
20             (Shape=1, Severity=0, Density=3)   (BI-RADS=4)  0.141519   
21             (Shape=2, Severity=0, Density=3)   (BI-RADS=4)  0.120708   
22             (BI-RADS=4, Severity=0, Shape=1)    (Margin=1)  0.142560   
23              (Shape=1, Severity=0, Margin=1)   (BI-RADS=4)  0.142560   
24              (Shape=2, Severity=0, Margin=1)   (BI-RADS=4)  0.108221   
25            (BI-RADS=5, Margin=4, Severity=1)   (Density=3)  0.118626   
26             (BI-RADS=5, Margin=4, Density=3)  (Severity=1)  0.118626   
27             (Shape=4, BI-RADS=5, Severity=1)   (Density=3)  0.221644   
28              (BI-RADS=5, Shape=4, Density=3)  (Severity=1)  0.221644   
29               (Shape=1, Margin=1, Density=3)  (Severity=0)  0.143600   
30             (Shape=1, Severity=0, Density=3)    (Margin=1)  0.143600   
31              (Shape=1, Severity=0, Margin=1)   (Density=3)  0.143600   
32              (Severity=1, Margin=4, Shape=4)   (Density=3)  0.133195   
33   (BI-RADS=4, Margin=1, Severity=0, Shape=1)   (Density=3)  0.133195   
34    (BI-RADS=4, Density=3, Margin=1, Shape=1)  (Severity=0)  0.133195   
35   (Density=3, Margin=1, Severity=0, Shape=1)   (BI-RADS=4)  0.133195   
36  (BI-RADS=4, Density=3, Severity=0, Shape=1)    (Margin=1)  0.133195   

    confidence  
0     0.938462  
1     0.933002  
2     0.900356  
3     0.918440  
4     0.924051  
5     0.906040  
6     0.933566  
7     0.911263  
8     0.920904  
9     0.962264  
10    0.955357  
11    0.924051  
12    0.929530  
13    0.943038  
14    0.905138  
15    0.927126  
16    0.932886  
17    0.908497  
18    0.912752  
19    0.931507  
20    0.931507  
21    0.920635  
22    0.938356  
23    0.919463  
24    0.928571  
25    0.904762  
26    0.904762  
27    0.930131  
28    0.914163  
29    0.901961  
30    0.945205  
31    0.926174  
32    0.907801  
33    0.934307  
34    0.920863  
35    0.927536  
36    0.941176

In [3]:
# exclude BI-RADS and predict the severity

severity_rule = rule[
    (rule['support'] >= 0.1) &
    (rule['consequents'].apply(lambda x: 'Severity=0' in x or 'Severity=1' in x)) &
    (~rule['antecedents'].apply(lambda x: any('BI-RADS' in item for item in x)))
]

severity_rule[["antecedents", "consequents", "support", "confidence"]]



antecedents   consequents  support  confidence
29  (Shape=1, Margin=1, Density=3)  (Severity=0)   0.1436    0.901961

<p style="font-size:20px;">When the Density of the lesion is 3, the Shape of it is round and the Margin of it is circumscribed, then the lesions are usually benign.
This rule has support of 14.36% in the dataset without BI-RADS. And it is correct in 90.20% of cases (90.20% Confidence). </p>

# Question 3
<p style="font-size:22px;">Find some rules proving that BI-RADS is not always precisely indicate benign or malign</p>

<p style="font-size:20px;">Reset the confidence threshold to 0.7. It is to ensure that the rules found have a high level of confidence, while increasing the number of potentially useful rules found for subsequent screening. This approach can explore the relationship between BI-RADS and Severity more extensively.</p>

In [4]:
from mlxtend.frequent_patterns import apriori, association_rules

# Generate association rules
rules_birads = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.7)

# Filter rules where 'BI-RADS' is part of the antecedents or consequents and relates to 'Severity'
birads_inaccuracy_rules = rules_birads[
    (rules_birads['antecedents'].apply(lambda x: any('BI-RADS' in item for item in x)) |
     rules_birads['consequents'].apply(lambda x: any('BI-RADS' in item for item in x))) &
    (rules_birads['consequents'].apply(lambda x: 'Severity=0' in x or 'Severity=1' in x))
]

birads_inaccuracy_rules[["antecedents", "consequents", "support", "confidence"]]



antecedents  \
7                                  (BI-RADS=5)   
29                      (BI-RADS=4, Density=3)   
38                       (BI-RADS=4, Margin=1)   
40                                  (Margin=1)   
46                        (BI-RADS=4, Shape=1)   
48                        (BI-RADS=4, Shape=2)   
52                      (BI-RADS=5, Density=3)   
54                                 (BI-RADS=5)   
57                       (BI-RADS=5, Margin=4)   
62                        (BI-RADS=5, Shape=4)   
92            (BI-RADS=4, Margin=1, Density=3)   
95                       (BI-RADS=4, Margin=1)   
96                       (Margin=1, Density=3)   
104            (BI-RADS=4, Shape=1, Density=3)   
107                       (BI-RADS=4, Shape=1)   
108                       (Shape=1, Density=3)   
111            (BI-RADS=4, Shape=2, Density=3)   
114                       (Shape=2, Density=3)   
115             (BI-RADS=4, Shape=1, Margin=1)   
118                       (BI-RADS=4, Shape=1)   
119                        (Shape=1, Margin=1)   
122             (BI-RADS=4, Shape=2, Margin=1)   
124                        (Shape=2, Margin=1)   
126           (BI-RADS=5, Margin=4, Density=3)   
127                      (BI-RADS=5, Margin=4)   
133            (BI-RADS=5, Shape=4, Density=3)   
136                       (BI-RADS=5, Shape=4)   
137                     (BI-RADS=5, Density=3)   
140             (BI-RADS=5, Margin=4, Shape=4)   
152  (BI-RADS=4, Density=3, Margin=1, Shape=1)   
155             (BI-RADS=4, Margin=1, Shape=1)   
157             (Density=3, Margin=1, Shape=1)   
159            (BI-RADS=4, Density=3, Shape=1)   
162                       (Density=3, Shape=1)   

                           consequents   support  confidence  
7                         (Severity=1)  0.297607    0.826590  
29                        (Severity=0)  0.340271    0.773050  
38                        (Severity=0)  0.269511    0.817035  
40             (BI-RADS=4, Severity=0)  0.269511    0.705722  
46                        (Severity=0)  0.151925    0.793478  
48                        (Severity=0)  0.140479    0.780347  
52                        (Severity=1)  0.277836    0.911263  
54             (Severity=1, Density=3)  0.277836    0.771676  
57                        (Severity=1)  0.131113    0.863014  
62                        (Severity=1)  0.238293    0.870722  
92                        (Severity=0)  0.238293    0.905138  
95             (Severity=0, Density=3)  0.238293    0.722397  
96             (BI-RADS=4, Severity=0)  0.238293    0.814947  
104                       (Severity=0)  0.141519    0.912752  
107            (Severity=0, Density=3)  0.141519    0.739130  
108            (BI-RADS=4, Severity=0)  0.141519    0.777143  
111                       (Severity=0)  0.120708    0.878788  
114            (BI-RADS=4, Severity=0)  0.120708    0.738854  
115                       (Severity=0)  0.142560    0.830303  
118             (Severity=0, Margin=1)  0.142560    0.744565  
119            (BI-RADS=4, Severity=0)  0.142560    0.740541  
122                       (Severity=0)  0.108221    0.812500  
124            (BI-RADS=4, Severity=0)  0.108221    0.707483  
126                       (Severity=1)  0.118626    0.904762  
127            (Severity=1, Density=3)  0.118626    0.780822  
133                       (Severity=1)  0.221644    0.914163  
136            (Severity=1, Density=3)  0.221644    0.809886  
137              (Shape=4, Severity=1)  0.221644    0.726962  
140                       (Severity=1)  0.105099    0.893805  
152                       (Severity=0)  0.133195    0.920863  
155            (Severity=0, Density=3)  0.133195    0.775758  
157            (BI-RADS=4, Severity=0)  0.133195    0.836601  
159             (Severity=0, Margin=1)  0.133195    0.859060  
162  (BI-RADS=4, Severity=0, Margin=1)  0.133195    0.731429

<p style="font-size:20px;">We have discovered many rules here, including BI-RADS and Severity, but still need to refine.</p>

In [5]:
birads_inaccuracy_rules2 = rules_birads[
    (rules_birads['antecedents'].apply(lambda x: any('BI-RADS' in item for item in x))) &
    (rules_birads['consequents'].apply(lambda x: 'Severity=1' in x))
]

birads_inaccuracy_rules2[["antecedents", "consequents", "support", "confidence"]]

antecedents              consequents   support  \
7                         (BI-RADS=5)             (Severity=1)  0.297607   
52             (BI-RADS=5, Density=3)             (Severity=1)  0.277836   
54                        (BI-RADS=5)  (Severity=1, Density=3)  0.277836   
57              (BI-RADS=5, Margin=4)             (Severity=1)  0.131113   
62               (BI-RADS=5, Shape=4)             (Severity=1)  0.238293   
126  (BI-RADS=5, Margin=4, Density=3)             (Severity=1)  0.118626   
127             (BI-RADS=5, Margin=4)  (Severity=1, Density=3)  0.118626   
133   (BI-RADS=5, Shape=4, Density=3)             (Severity=1)  0.221644   
136              (BI-RADS=5, Shape=4)  (Severity=1, Density=3)  0.221644   
137            (BI-RADS=5, Density=3)    (Shape=4, Severity=1)  0.221644   
140    (BI-RADS=5, Margin=4, Shape=4)             (Severity=1)  0.105099   

     confidence  
7      0.826590  
52     0.911263  
54     0.771676  
57     0.863014  
62     0.870722  
126    0.904762  
127    0.780822  
133    0.914163  
136    0.809886  
137    0.726962  
140    0.893805

<p style="font-size:20px;">Here, we find that the antecedents containing BI-RADS that can derive Severity=1 are all "BI-RADS=5". The high value of BI-RADS indicates a high possibility of malignancy, which is consistent with the consequents terms. So need to mining deeper.</p>

In [6]:
birads_inaccuracy_rules3 = rules_birads[
    (rules_birads['antecedents'].apply(lambda x: any('BI-RADS' in item for item in x))) &
    (rules_birads['consequents'].apply(lambda x: 'Severity=0' in x))
]

birads_inaccuracy_rules3[["antecedents", "consequents", "support", "confidence"]]

antecedents              consequents  \
29                      (BI-RADS=4, Density=3)             (Severity=0)   
38                       (BI-RADS=4, Margin=1)             (Severity=0)   
46                        (BI-RADS=4, Shape=1)             (Severity=0)   
48                        (BI-RADS=4, Shape=2)             (Severity=0)   
92            (BI-RADS=4, Margin=1, Density=3)             (Severity=0)   
95                       (BI-RADS=4, Margin=1)  (Severity=0, Density=3)   
104            (BI-RADS=4, Shape=1, Density=3)             (Severity=0)   
107                       (BI-RADS=4, Shape=1)  (Severity=0, Density=3)   
111            (BI-RADS=4, Shape=2, Density=3)             (Severity=0)   
115             (BI-RADS=4, Shape=1, Margin=1)             (Severity=0)   
118                       (BI-RADS=4, Shape=1)   (Severity=0, Margin=1)   
122             (BI-RADS=4, Shape=2, Margin=1)             (Severity=0)   
152  (BI-RADS=4, Density=3, Margin=1, Shape=1)             (Severity=0)   
155             (BI-RADS=4, Margin=1, Shape=1)  (Severity=0, Density=3)   
159            (BI-RADS=4, Density=3, Shape=1)   (Severity=0, Margin=1)   

      support  confidence  
29   0.340271    0.773050  
38   0.269511    0.817035  
46   0.151925    0.793478  
48   0.140479    0.780347  
92   0.238293    0.905138  
95   0.238293    0.722397  
104  0.141519    0.912752  
107  0.141519    0.739130  
111  0.120708    0.878788  
115  0.142560    0.830303  
118  0.142560    0.744565  
122  0.108221    0.812500  
152  0.133195    0.920863  
155  0.133195    0.775758  
159  0.133195    0.859060

<p style="font-size:20px;">Here, we search for the rules that have BI-RADS in its antecedents and Severity=0 in its consequents. Then we find something interesting.</p>

In [7]:

specific_rule = rules_birads[
    (rules_birads['antecedents'].apply(lambda x: 'BI-RADS=4' in x)) &
    (rules_birads['consequents'].apply(lambda x: 'Severity=0' in x))
]


specific_rule[["antecedents", "consequents", "support", "confidence"]]

antecedents              consequents  \
29                      (BI-RADS=4, Density=3)             (Severity=0)   
38                       (BI-RADS=4, Margin=1)             (Severity=0)   
46                        (BI-RADS=4, Shape=1)             (Severity=0)   
48                        (BI-RADS=4, Shape=2)             (Severity=0)   
92            (BI-RADS=4, Margin=1, Density=3)             (Severity=0)   
95                       (BI-RADS=4, Margin=1)  (Severity=0, Density=3)   
104            (BI-RADS=4, Shape=1, Density=3)             (Severity=0)   
107                       (BI-RADS=4, Shape=1)  (Severity=0, Density=3)   
111            (BI-RADS=4, Shape=2, Density=3)             (Severity=0)   
115             (BI-RADS=4, Shape=1, Margin=1)             (Severity=0)   
118                       (BI-RADS=4, Shape=1)   (Severity=0, Margin=1)   
122             (BI-RADS=4, Shape=2, Margin=1)             (Severity=0)   
152  (BI-RADS=4, Density=3, Margin=1, Shape=1)             (Severity=0)   
155             (BI-RADS=4, Margin=1, Shape=1)  (Severity=0, Density=3)   
159            (BI-RADS=4, Density=3, Shape=1)   (Severity=0, Margin=1)   

      support  confidence  
29   0.340271    0.773050  
38   0.269511    0.817035  
46   0.151925    0.793478  
48   0.140479    0.780347  
92   0.238293    0.905138  
95   0.238293    0.722397  
104  0.141519    0.912752  
107  0.141519    0.739130  
111  0.120708    0.878788  
115  0.142560    0.830303  
118  0.142560    0.744565  
122  0.108221    0.812500  
152  0.133195    0.920863  
155  0.133195    0.775758  
159  0.133195    0.859060

<p style="font-size:20px;">
So, We have a new finding. When BI-RADS=4, which means a slightly higher possibility of malignancy, the Severity is 0. It contradicts the description of BI-RADS. Normally, when BI-RADS has a high value, Severity tends to be more malignant.<br>
Even the confidence of the rule (Density=3, Margin=1, BI-RADS=4) ->(Severity=0) has reached 90%.
</p>

# Question 4
<p style="font-size:22px;">Find the rules and its support and confidence: Age=35 -> Severity=0.</p>

In [8]:
antecedent = df['Age=35']
consequent = df['Severity=0']

support = ((antecedent & consequent).sum()) / len(df)

confidence = ((antecedent & consequent).sum()) / antecedent.sum()

print(f"Support: {support:.6f}, Confidence: {confidence:.6f}")

Support: 0.010406, Confidence: 0.769231


<p style="font-size:20px;">
We find the rules: Age=35 -> Severity=0.<br><br>
It's Support: 0.010406.<br>
It's Confidence: 0.769231.<br><br>
This meas even though the confidence indicates that in the case of Age=35, most instances are benign, there are only 1.04% of the instances in the entire dataset meet the criteria of Age=35 and Severity=0. This value is really really low, indicating that this rule appears very infrequently in the dataset.<br><br>
So, I think this rule cannot bring us much value because its support is simply too low. There are not enough examples to prove it.
</p>

# Question 5
<p style="font-size:20px;">
Convert the "Age" to Binary and Mining some extra rules like:<br><br>
Age >=n, A1 =a1, ..., Ak =ak -> Severity=1
</p>

<p style="font-size:20px;">
I have an idea for handling missing values in 'Age'. Namely, mining age-related rules and filling them with the highest confidence level rules.
</p>

In [29]:
import math
import pandas as pd



data_missing = pd.read_csv(r"mammographic_masses.csv",delimiter=",",header=0) # The actual dataset with missing values Not be DELETED 

data_missing.replace('?', pd.NA, inplace=True)

data_missing


BI-RADS Age Shape Margin Density  Severity
0         5  67     3      5       3         1
1         4  43     1      1    <NA>         1
2         5  58     4      5       3         1
3         4  28     1      1       3         0
4         5  74     1      5    <NA>         1
..      ...  ..   ...    ...     ...       ...
956       4  47     2      1       3         0
957       4  56     4      5       3         1
958       4  64     4      5       3         0
959       5  66     4      5       3         1
960       4  62     3      3       3         0

[961 rows x 6 columns]

In [30]:
data_missing['Age'] = pd.to_numeric(data_missing['Age'], errors='coerce')
data_missing

BI-RADS   Age Shape Margin Density  Severity
0         5  67.0     3      5       3         1
1         4  43.0     1      1    <NA>         1
2         5  58.0     4      5       3         1
3         4  28.0     1      1       3         0
4         5  74.0     1      5    <NA>         1
..      ...   ...   ...    ...     ...       ...
956       4  47.0     2      1       3         0
957       4  56.0     4      5       3         1
958       4  64.0     4      5       3         0
959       5  66.0     4      5       3         1
960       4  62.0     3      3       3         0

[961 rows x 6 columns]

In [31]:
rules = association_rules(frequent_itemsets, metric="confidence")

age_rules = rules[rules['consequents'].apply(lambda x: any('Age' in item for item in x))]
age_rules

Empty DataFrame
Columns: [antecedents, consequents, antecedent support, consequent support, support, confidence, lift, leverage, conviction, zhangs_metric]
Index: []

<p style="font-size:20px;">
There is no rules related to 'Age'. But why?
</p>

In [32]:
age_counts = data_missing['Age'].dropna().astype(int).value_counts()
most_frequent_age = age_counts.idxmax()
most_frequent_age_count = age_counts.max()

most_frequent_age, most_frequent_age_count

(np.int64(59), np.int64(36))

<p style="font-size:20px;">
Here I understand the reason, The most frequent Age is 59, but its support is only 36, which is 36 / 961 ≈ 0.037, which is much smaller than 0.1. So NO Age will be the candidates of the frequent items. So My idea may not be feasible.<br><br>
So, I chose to use median to fill in the missing values. Because it is robust and less sensitive to extreme values
</p>

In [35]:
median_age = data_missing['Age'].median()

# use median to fill the NA
data_missing['Age'].fillna(median_age, inplace=True)

print(data_missing['Age'].isna().sum())  

0


C:\Users\19928\AppData\Local\Temp\ipykernel_28772\4029746388.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data_missing['Age'].fillna(median_age, inplace=True)


<p style="font-size:20px;">
Then check the maligancy rate with different ages.
</p>

In [36]:
age_severity_counts = data_missing.groupby(['Age', 'Severity']).size().unstack(fill_value=0)

# rename
age_severity_counts.columns = ['Benign_Count', 'Malignant_Count']

# The rate
age_severity_counts['Total'] = age_severity_counts['Benign_Count'] + age_severity_counts['Malignant_Count']
age_severity_counts['Malignancy_Rate'] = age_severity_counts['Malignant_Count'] / age_severity_counts['Total']

# index the Age
age_severity_counts.index.name = "Age"

age_severity_counts[['Benign_Count', 'Malignant_Count', 'Total', 'Malignancy_Rate']]


Benign_Count  Malignant_Count  Total  Malignancy_Rate
Age                                                        
18.0             1                0      1              0.0
19.0             4                0      4              0.0
20.0             1                0      1              0.0
21.0             6                0      6              0.0
22.0             3                0      3              0.0
...            ...              ...    ...              ...
86.0             1                4      5              0.8
87.0             0                5      5              1.0
88.0             0                1      1              1.0
93.0             0                1      1              1.0
96.0             0                1      1              1.0

[73 rows x 4 columns]

In [44]:
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, association_rules

n = 60

data_missing['Age_binary'] = data_missing['Age'].apply(lambda x: 1 if x >= n else 0)


data_list = data_missing.values.tolist()

# dealing with tht missing values exclude 'Age'
for i in range(len(data_list)):
    j = 0
    while j < len(data_list[i]):
        # is NAN?
        if pd.isna(data_list[i][j]) and data_missing.columns[j] != 'Age':  
            del data_list[i][j]  # del
            j -= 1
        j += 1

# add clo name to mining the rules
for i in range(len(data_list)):
    for j in range(len(data_list[i])):
        data_list[i][j] = data_missing.columns[j] + "=" + str(data_list[i][j])


te_age = TransactionEncoder()
te_array_age = te_age.fit(data_list).transform(data_list)
df_age = pd.DataFrame(te_array_age, columns=te_age.columns_)


freq_itemsets = apriori(df_age, min_support=0.1, use_colnames=True)
ar_ages = association_rules(freq_itemsets, metric="confidence", min_threshold=0.9)


ar_ages[["antecedents", "consequents", "support", "confidence"]]


antecedents             consequents  \
0                                      (Age_binary=1)             (Density=3)   
1                                          (Margin=5)             (Density=3)   
2                           (BI-RADS=4, Age_binary=0)             (Density=3)   
3                          (Age_binary=0, Severity=1)             (Density=3)   
4                             (Age_binary=0, Shape=1)             (Density=3)   
..                                                ...                     ...   
88  (Shape=1, Age_binary=0, Margin=1, BI-RADS=4, D...            (Severity=0)   
89  (Shape=1, Age_binary=0, Margin=1, Severity=0, ...             (BI-RADS=4)   
90  (Shape=1, Age_binary=0, BI-RADS=4, Severity=0,...              (Margin=1)   
91     (BI-RADS=4, Age_binary=0, Severity=0, Shape=1)   (Density=3, Margin=1)   
92      (BI-RADS=4, Age_binary=0, Density=3, Shape=1)  (Severity=0, Margin=1)   

     support  confidence  
0   0.335068    0.930636  
1   0.127992    0.938931  
2   0.309053    0.900000  
3   0.152966    0.907407  
4   0.138398    0.936620  
..       ...         ...  
88  0.112383    0.947368  
89  0.112383    0.923077  
90  0.112383    0.964286  
91  0.112383    0.907563  
92  0.112383    0.915254  

[93 rows x 4 columns]

In [46]:
final_rules = ar_ages[
    ar_ages['antecedents'].apply(lambda x: any('Age_binary=1' in item for item in x)) &
    ar_ages['consequents'].apply(lambda x: any('Severity=1' in item for item in x))
]

final_rules[['antecedents', 'consequents', 'support', 'confidence']]

antecedents              consequents  \
10                      (Age_binary=1, BI-RADS=5)             (Severity=1)   
41           (Age_binary=1, BI-RADS=5, Density=3)             (Severity=1)   
43             (Age_binary=1, BI-RADS=5, Shape=4)             (Severity=1)   
82  (Age_binary=1, BI-RADS=5, Shape=4, Density=3)             (Severity=1)   
83             (Age_binary=1, BI-RADS=5, Shape=4)  (Severity=1, Density=3)   

     support  confidence  
10  0.188345    0.932990  
41  0.180021    0.945355  
43  0.149844    0.947368  
82  0.144641    0.958621  
83  0.144641    0.914474

<p style="font-size:20px;">
So what is displayed here are all the rules that meet the conditions. That is: <br><br>
Age >=60, BI-RADS=5 -> Severity=1<br>
Age >=60, BI-RADS=5, Density=3 -> Severity=1<br>
Age >=60, BI-RADS=5, Shape=4 -> Severity=1<br>
Age >=60, BI-RADS=5, Shape=4, Density=3 -> Severity=1<br>
Age >=60, BI-RADS=5, Shape=4 -> Severity=1, Density=3<br>
</p>